### DISCUSSION TITLE
[REQUEST] Enable flush method from dynamoDB cache

### DISCUSSION THREAD

**Discussion by didix16** [laravel/framework#37855] [Ideas]
Hi everyone,

I'm using Laravel Vapor for a serverless project and also I'm using DynamoDB as a caching system. I just realized that flush method is throwing and Exception:

```php
/**
     * Remove all items from the cache.
     *
     * @return bool
     *
     * @throws \RuntimeException
     */
    public function flush()
    {
        throw new RuntimeException('DynamoDb does not support flushing an entire table. Please create a new table.');
    }
```

I don't understand why because it can be implemented by using the DynamoDbClient::scan method and then delete all items looping through them or just deleting the table and recreating it again.

Here is my approach:

```php

 // @method \Aws\Result scan(array $args = []) <== This is from DynamoDBClient class
/**
     * Remove all items from the cache.
     *
     * @return bool
     */
    public function flush()
    {
        // This returns the entire items list on dynamoDB table
        $items = $this->dynamo->scan([
            'TableName' => $this->table,
            ...what ever the options this API call needs ....
        ]);
        $result = true;
        
        // I dont know how the scan method result  and item structure are but suppose it has a getKey() method to get the key....
        foreach($items as $item){
           $result =  $this->forget($item->getKey()) && $result;
        }
        // if all items could be deleted then flush will return true, else false.
        return $result;
    }
```

So what do you think? Hope could be implemented.

Thanks in advance.

**Comment by sebastiaanluca** [comment]
Running into the same issue. Have some data in cache in DynamoDB on Vapor, but I can't seem to clear it. What's the best approach here?

**Reply by sebastiaanluca** [reply]
For those interested, as a workaround I did the following (needed to flush the entire cache on each deployment):

Ended up with writing the git commit hash to a file on build in vapor.yml (using a command), then using that hash in the cache prefix config setting (so it ends up being something like `202hkl2jh34xg23kjh42j4k234j2kg_app_name_cache`). Additionally, set my max caching for any value written to the cache to a week or so, so outdated DynamoDB entries are never kept around for long.

Now my cache is "cleared" on each deployment and the old caches are removed max a week from now.

Alternatively, either you scan the whole table and remove each record one by one (takes a while and costly), or keep track of each cache key and remove it manually during deployment. Also looked into creating a new DynamoDB table, but Vapor configures it according to their settings, so no way of having these exactly the same.

**Reply by oliworx** [reply]
> Ended up with writing the git commit hash to a file on build in vapor.yml (using a command), then using that hash in the cache prefix config setting (so it ends up being something like 202hkl2jh34xg23kjh42j4k234j2kg_app_name_cache). Additionally, set my max caching for any value written to the cache to a week or so, so outdated DynamoDB entries are never kept around for long.

You could simply use VAPOR_COMMIT_HASH for the cache prefix

**Comment by kirkkhua** [comment]
Have the same problem when running `php artisan cache:clear` on Laravel Framework 9.9.0

**Comment by nikolay484** [comment]
@didix16 
a little update for your approach 
```
/**
     * Remove all items from the cache.
     *
     * @return bool
     *
     * @throws \RuntimeException
     */
    public function flush()
    {
        $items = $this->dynamo->scan([
            'TableName' => $this->table,
//            ...what ever the options this API call needs ....
        ]);
        $result = true;
        $i = 0;
        while($i < count($items['Items'])) {
            $result =  $this->forget(preg_replace('/.+:/', '', $items['Items'][$i]['key']['S'])) && $result;
            $i++;
        }
        return $result;
    }
```

**Comment by nikolay484** [comment]
Also Added option to override the command without editing vendor files. 
I added to AppServiceProvider to a register() function 
```
$this->app->extend(ClearCommand::class, function ($data) {
            if(env('CACHE_DRIVER') == 'dynamodb') {
                $config = [
                    'driver' => 'dynamodb',
                    'credentials' => array(
                        'key' => env('CACHE_ACCESS_KEY'),
                        'secret' => env('CACHE_SECRET_ACCESS_KEY'),
                    ),
                    'region' => env('CACHE_DEFAULT_REGION'),
                    'table' => env('CACHE_DYNAMODB_TABLE'),
                    'version' => 'latest'
                ];
                $client = new \Aws\DynamoDb\DynamoDbClient($config);
                $dynamoDbStore = new \Illuminate\Cache\DynamoDbStore($client, env('CACHE_DYNAMODB_TABLE'));

                $items = $client->scan([
                    'TableName' => $config['table'],
//            ...what ever the options this API call needs ....
                ]);
                $result = true;
                $i = 0;
                while($i < count($items['Items'])) { // i used while to not exceed the memory of php

                    $result =  $dynamoDbStore->forget(preg_replace('/.+:/', '', $items['Items'][$i]['key']['S'])) && $result;
                    $i++;
                }
            }

        });
```

**Comment by bradley-r-martin** [comment]
Thanks for the solution, This worked for me however i had to override the entire DynamoDbStore.php file to get this to work, i achieved this using composer exclude-from-classmap

`  "exclude-from-classmap": [
            "vendor/laravel/framework/src/Illuminate/Cache/DynamoDbStore.php"
        ],
        "files": [
            "./app/Overrides/DynamoDbStore.php"
        ],`

Would be great if there was a way to just override the flush method on the class instead.
